apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -845,124 +658,201 @@ do_ensure_all_iterators_closed(_DSSessionID) ->
     ok.
 
 %%--------------------------------------------------------------------
-%% Reading batches
+%% Normal replay:
 %%--------------------------------------------------------------------
 
--spec renew_streams(session()) -> ok.
-renew_streams(#{id := SessionId, subscriptions := Subscriptions}) ->
-    transaction(fun() ->
-        ExistingStreams = mnesia:read(?SESSION_STREAM_TAB, SessionId, write),
-        subs_fold(
-            fun(TopicFilter, #{start_time := StartTime}, Streams) ->
-                TopicFilterWords = emqx_topic:words(TopicFilter),
-                renew_topic_streams(SessionId, TopicFilterWords, StartTime, Streams)
-            end,
-            ExistingStreams,
-            Subscriptions
-        )
-    end),
-    ok.
+fetch_new_messages(Session = #{s := S}, ClientInfo) ->
+    Streams = emqx_persistent_session_ds_stream_scheduler:find_new_streams(S),
```

Comment:
While thinking about consistency soundness of those changes, it just occurred to me that this is unsafe because `renew_streams/1` is nondeterministic. If the session fails to be persisted then the client will very likely see different order of messages during reconnect and replay.

Reply:
There are three evils:
1) Transactions (with a singleton locker process handling everything in the node)
2) Storing everything in one huge record, and receiving comments about write amplification :wink: 
3) Async dirty with flimsy guarantees.

I chose 3 mostly because mnesia here is as temporary as it gets.

Reply:
Yeah. 🫠 Kinda feel the pain.

> because mnesia here is as temporary as it gets.

Then why not choose 1 instead? Could at least free ourselves from thinking about those edge cases. Now it feels like an unsafe optimization that looks sort of odd as part of a refactoring effort.

Reply:
Heh, both you and I probably missed the thread. This one is about nondeterministic replays.

Reply:
> But I think LTS may lead to independent streams containing the same topic (the paths before and after "learning")?

This must never happen within a given generation: `get_streams` for the LTS must always return the _original_ index for the topic.
New topics will end up in the wildcard-augmented stream, but the old ones are stored in their own streams.

Reply:
> From the [EIP](https://github.com/emqx/eip/blob/162edbbbfcc6dce1eed5f164d516cb9e7214330e/active/0023-rocksdb-message-persistence.md?plain=1#L24), we don't provide guarantees about ordering between different topics.

Just wanted to clarify. The problem lies in how existing sessions are replayed in the event of a failure, when recent session state updates were lost. The client will reconnect and see that the session was present, but then receive some of the messages it has already seen but with different PacketIDs. This essentially violates [MQTT spec](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901238), and makes it harder for clients to process messages idempotently.